pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java

```diff
@@ -0,0 +1,383 @@
+                } else {
+                    // Done replaying
+                    future.complete(null);
+                }
+            }
+
+            @Override
+            public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
+                future.completeExceptionally(exception);
+            }
+        }, null);
+    }
+
+    public CompletableFuture<Void> initialize() {
+        // Check whether the dedup cursor was already present
+        for (ManagedCursor cursor : managedLedger.getCursors()) {
+            if (cursor.getName().equals(PersistentTopic.DEDUPLICATION_CURSOR_NAME)) {
+                // Deduplication was enabled before
+                this.status = Status.Recovering;
+                this.managedCursor = cursor;
+                break;
+            }
+        }
+
+        if (status == Status.Recovering) {
+            // Recover the current cursor and then check the configuration
+            return recoverSequenceIdsMap().thenCompose(v -> checkStatus());
+        } else {
+            // No-op
+            return CompletableFuture.completedFuture(null);
+        }
+    }
+
+    /**
+     * Check the status of deduplication. If the configuration has changed, it will enable/disable deduplication,
+     * returning a future to track the completion of the task
+     */
+    public CompletableFuture<Void> checkStatus() {
+        if (status == Status.Recovering || status == Status.Removing) {
+            // If there's already a transition happening, check later for status
+            pulsar.getExecutor().schedule(this::checkStatus, 1, TimeUnit.MINUTES);
+            return CompletableFuture.completedFuture(null);
+        }
+
+        return isDeduplicationEnabled().thenCompose(shouldBeEnabled -> {
+            if (status != Status.Disabled && !shouldBeEnabled) {
+                // Disabled deduping
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                status = Status.Removing;
+                managedLedger.asyncDeleteCursor(PersistentTopic.DEDUPLICATION_CURSOR_NAME, new DeleteCursorCallback() {
+
+                    @Override
+                    public void deleteCursorComplete(Object ctx) {
+                        status = Status.Disabled;
+                        managedCursor = null;
+                        highestSequencedPushed.clear();
+                        highestSequencedPersisted.clear();
+                        future.complete(null);
+                        log.info("[{}] Disabled deduplication", topic.getName());
+                    }
+
+                    @Override
+                    public void deleteCursorFailed(ManagedLedgerException exception, Object ctx) {
+                        log.warn("[{}] Failed to disable deduplication: {}", topic.getName(), exception.getMessage());
+                        status = Status.Failed;
+                        future.completeExceptionally(exception);
```

Comment:
is it a `ledger leak`? should we retry X time?

Reply:
It's not a ledger leak since it's the delete *cursor* operation that failed. In this case the cursor will stay there. I'll make sure to schedule a check in this case to retry to disable deduplication.